pkg/curatedpackages/bundleclient.go

```diff
@@ -0,0 +1,172 @@
+package curatedpackages
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	packagesv1 "github.com/aws/eks-anywhere-packages/api/v1alpha1"
+	"github.com/aws/eks-anywhere-packages/pkg/artifacts"
+	"github.com/aws/eks-anywhere-packages/pkg/bundle"
+	"github.com/aws/eks-anywhere/pkg/constants"
+)
+
+// BundleClient abstracts the source of package bundle information.
+//
+// Whether the source is a registry, a cluster, the BundleClient should
+// abstract the details of retrieving a particular curated packages bundle.
+type BundleClient interface {
+	ActiveOrLatest(ctx context.Context) (*packagesv1.PackageBundle, error)
+}
+
+// BundleClientOptions alter the behavior of the bundle client when retrieving
+// package bundle information.
+type BundleClientOptions struct {
+	// ClusterName identifies the cluster with which to communicate.
+	ClusterName string
+	// KubeConfig specifies a kubeconfig file to use.
+	//
+	// Defaults to the KUBECONFIG envrionment variable, and can fall back to
+	// ~/.kube/config or similar.
+	KubeConfig string
+	// KubeVersion specifies a kubernetes version requirement for package bundles.
+	//
+	// Required if consulting a registry for package bundle information.
+	KubeVersion string
+	// Registry as an OCI Registry URL.
+	//
+	// Defaults to the EKS-A ECR. Used when consulting a registry for package
+	// bundle information.
+	Registry string
+	// RegistryClient will be used if supplied.
+	//
+	// Specifying a custom registry client is optional, but could be needed
+	// when specifying a custom registry mirror that requires authentication.
+	RegistryClient bundle.RegistryClient
+	// restConfigurator is used to inject testing constructs.
+	//
+	// It shouldn't be required for normal use.
+	restConfigurator restConfigurator
+}
+
+// NewBundleClient constructs a BundleClient based on the value of source.
+func NewBundleClient(source BundleSource, opts BundleClientOptions) (BundleClient, error) {
```

Comment:
Having that `source` argument makes me think this should probably be two constructors and the caller should decide which one to call.

Reply:
I could, but I think this is a better solution because there are a few different CLI commands that have to instantiate a bundle client with these same options. NewBundleClient lifts that common logic out of the individual CLI commands to a single, more authoritative place.

Reply:
Maybe I need more context to see how this is going to be used by consumers.
Should we table this conversation until we move to those PRs? As long as you are OK revisiting this discussion when the time comes.

Reply:
I'm willing to re-open it later, but the context you need is in the code today. Look at the commands for list packages, generate package, or install package.